rpcs3/Emu/Cell/Modules/cellSearch.cpp

```diff
@@ -76,13 +138,39 @@ error_code cellSearchInitialize(CellSearchMode mode, u32 container, vm::ptr<Cell
 
 error_code cellSearchFinalize()
 {
-	cellSearch.warning("cellSearchFinalize()");
+	cellSearch.todo("cellSearchFinalize()");
 
 	const auto search = g_fxo->get<search_info>();
 
+	switch(search->state)
+	{
+		case SearchState::IDLE:
+			break;
+		case SearchState::NOT_INITIALIZED:
+			return CELL_SEARCH_ERROR_NOT_INITIALIZED;
+		case SearchState::FINALIZING:
+			return CELL_SEARCH_ERROR_FINALIZING;
+		case SearchState::IN_PROGRESS:
+		case SearchState::INITIALIZING:
+		case SearchState::CANCELING:
+			return CELL_SEARCH_ERROR_BUSY;
+		default:
+			return CELL_SEARCH_ERROR_GENERIC;
+	}
+
+	search->state = SearchState::FINALIZING;
+
+	// Umount hard links
+	for (auto& link : search->content_links)
+	{
+		vfs::mount(link.first, "");
+	}
+
 	sysutil_register_cb([=](ppu_thread& ppu) -> s32
 	{
-		search->func(ppu, CELL_SEARCH_EVENT_FINALIZE_RESULT, CELL_OK, vm::null, search->userData);
+		search->content_links.clear();
```

Comment:
Here is a better example where thread safety is important.

Reply:
What do you suggest? `shared_lock` around those calls? If that is a concern I'd make a `search->getState()` and `search->setState(x)` so it's easier to manage and not forget.

Reply:
Yes, sounds good.